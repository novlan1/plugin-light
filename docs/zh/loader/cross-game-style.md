# cross-game-style

`scss`链接的关键词编译，替换关键词`@TIP_STYLE_NAME`。


## 如何使用

1. 先安装 npm 包：

```bash
npm i uni-plugin-light -D
```

2. 在 `vue.config.js` 中添加如下设置：

```js
const CROSS_GAME_STYLE_LOADER = 'uni-plugin-light/lib/loader/cross-game-style';

module.export = {
  chainWebpack(config) {
    config.module
      .rule('vue')
      .test(/\.vue$/)
      .use(CROSS_GAME_STYLE_LOADER) 
      // 处理样式的loader，必须在vue-loader前执行
      .loader(CROSS_GAME_STYLE_LOADER)
      .end();
  }
}
```


## loader参数

| 参数      | 说明                 | 类型              | 默认值    |
| --------- | -------------------- | ----------------- | --------- |
| styleName | 样式文件名，不含后缀 | _string \| array_ | -         |
| platforms | 处理的平台，默认全部 | _Array\<string\>_ | `['ALL']` |
