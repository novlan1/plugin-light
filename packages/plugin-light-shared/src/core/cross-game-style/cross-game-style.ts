import * as fs from 'fs';
import { getStyleName } from './style-name';
import { TIP_STYLE_NAME } from './config';
import type { ICrossGameStyleOptions } from './types';


function tryRemoveImport(source: string, removeImport = false) {
  let res = source;
  if (removeImport) {
    res = res.replace(`src="${TIP_STYLE_NAME}"`, '').replace(`src='${TIP_STYLE_NAME}'`, '');
  } else {
    res = res.replace(TIP_STYLE_NAME, '');
  }
  return res;
}


export function crossGameStyle({
  source,
  options,
  dir,

  removeImport = false,
}: {
  source: string;
  options?: ICrossGameStyleOptions;
  dir: string;
  removeImport?: boolean;
}) {
  if (!source.includes(TIP_STYLE_NAME)) {
    return source;
  }

  let styleName: String | Array<String> = '';
  // 使用 env.local 的样式 VUE_APP_DIR = module/ingame-nba，即为 nba
  if (options?.styleName) {
    styleName = options.styleName;
  } else if (getStyleName()) {
    styleName = getStyleName();
  }

  if (Array.isArray(styleName)) {
    if (styleName.length > 1) {
      styleName = styleName.filter((item) => {
        const cssAbsolutePath = `${dir}/css/${item}.scss`;
        return fs.existsSync(cssAbsolutePath);
      });

      if (styleName.length > 1) {
        const styleTags = styleName
          .filter((item) => {
            const cssAbsolutePath = `${dir}/css/${item}.scss`;
            return removeImport ? fs.existsSync(cssAbsolutePath) : true;
          })
          .map(item => `.${item} {@import './css/${item}.scss';}`);

        const res = tryRemoveImport(source, removeImport);
        return res.replace(/<\/style>/, `</style>${['<style scoped lang="scss">', ...styleTags, '</style>'].join('')}`);
      }
    }
    styleName = styleName[0] || '';
  }

  const cssPath = `./css/${styleName}.scss`;
  const cssAbsolutePath = `${dir}/css/${styleName}.scss`;
  const exist = fs.existsSync(cssAbsolutePath);

  if (exist) {
    return source.replace(TIP_STYLE_NAME, cssPath);
  }

  return tryRemoveImport(source, removeImport);
}
