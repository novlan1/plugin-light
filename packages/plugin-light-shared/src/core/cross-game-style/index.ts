export { crossGameStyle } from './cross-game-style';
export type { ICrossGameStyleOptions } from './types';
