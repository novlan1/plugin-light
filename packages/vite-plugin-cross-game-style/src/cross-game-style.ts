import path from 'path';
import type { ICrossGameStyleOptions } from './types';
import { crossGameStyle } from 'plugin-light-shared';

function genWarn(source: string, id: string, warnList: Array<string> = []) {
  warnList.forEach((item) => {
    if (source.includes(item)) {
      console.log(`[WARN] ${id} 文件中包含 ${item}`);
    }
  });
}


/**
 * Vite 插件 - 将多游戏样式文件占位符 `@TIP_STYLE_NAME` 转为 对应的文件路径
 * @param options.styleName - 样式名称
 */
export function crossGameStyleVitePlugin(options?: ICrossGameStyleOptions) {
  return {
    name: 'vite-plugin-cross-game-style',
    enforce: 'pre',
    transform(source: string, id: string) {
      genWarn(source, id, options?.warnList);

      if (!id.endsWith('.vue')) {
        return {
          code: source,
          map: null,
        };
      }

      const result = crossGameStyle({
        source,
        options,
        dir: path.dirname(id),
        removeImport: true,
      });

      return {
        code: result,
        map: null,
      };
    },
  };
}

