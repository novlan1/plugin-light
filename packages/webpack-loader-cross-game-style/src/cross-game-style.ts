import { getOptions } from 'loader-utils';
import { ALL_PLATFORM, crossGameStyle, type ICrossGameStyleOptions } from 'plugin-light-shared';
import { shouldUseLoader } from 'plugin-light-shared-vue2';


export function crossGameStyleLoader(this: any, source: string) {
  if (!shouldUseLoader.call(this, [ALL_PLATFORM])) return source;

  // 改为异步loader
  const callback = this.async();
  const options: ICrossGameStyleOptions = getOptions(this);
  const result = crossGameStyle({
    source,
    options,
    dir: this.context,
  });
  callback(null, result);
}

export {
  crossGameStyle,
  ICrossGameStyleOptions,
};
