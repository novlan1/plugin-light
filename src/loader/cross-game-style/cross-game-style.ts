import * as fs from 'fs';
import { getStyleName } from './style-name';
import { getOptions } from 'loader-utils';
import { TIP_STYLE_NAME } from './config';
import { shouldUseLoader, ALL_PLATFORM } from '../../helper/loader-options';


export function crossGameStyleLoader(this: any, source: string) {
  if (!shouldUseLoader.call(this, [ALL_PLATFORM])) return source;

  // 改为异步loader
  const callback = this.async();
  const options = getOptions(this);
  if (source.indexOf(TIP_STYLE_NAME) !== -1) {
    let styleName: String|Array<String> = '';
    // 使用env.local的样式 VUE_APP_DIR = module/ingame-nba，即为nba
    if (options?.styleName) {
      // @ts-ignore
      styleName = options.styleName;
    } else if (getStyleName()) {
      styleName = getStyleName();
    }
    if (Array.isArray(styleName)) {
      if (styleName.length > 1) {
        styleName = styleName.filter((item) => {
          const cssAbsolutePath = `${this.context}/css/${item}.scss`;
          return fs.existsSync(cssAbsolutePath);
        });
        if (styleName.length > 1) {
          const styleTags = styleName.map(item => `.${item} {@import './css/${item}.scss';}`);
          callback(null, source.replace(TIP_STYLE_NAME, '').
            replace(/<\/style>/, `</style>${['<style scoped lang="scss">', ...styleTags, '</style>'].join('')}`));
          return;
        }
      }
      styleName = styleName[0] || '';
    }
    const cssPath = `./css/${styleName}.scss`;
    const cssAbsolutePath = `${this.context}/css/${styleName}.scss`;
    const exist = fs.existsSync(cssAbsolutePath);
    if (exist) {
      callback(null, source.replace(TIP_STYLE_NAME, cssPath));
    } else {
      callback(null, source.replace(TIP_STYLE_NAME, ''));
    }
  } else {
    callback(null, source);
  }
}

