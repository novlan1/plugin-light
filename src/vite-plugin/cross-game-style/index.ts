export { crossGameStyleVitePlugin } from './cross-game-style';
