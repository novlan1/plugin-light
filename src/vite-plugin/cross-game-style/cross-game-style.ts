import path from 'path';
import fs from 'fs';
import { TIP_STYLE_NAME } from '../../webpack-loader/cross-game-style/config';
import type { ICrossGameStyleOptions } from './types';


/**
 * Vite 插件 - 将多游戏样式文件占位符 `@TIP_STYLE_NAME` 转为 对应的文件路径
 * @param options.styleName - 样式名称
 */
export function crossGameStyleVitePlugin(options?: ICrossGameStyleOptions) {
  const { styleName = '' } = options || {};

  return {
    name: 'vite-plugin-cross-game-style',

    transform(source: string, id: string) {
      let res = source;

      if (res.indexOf(TIP_STYLE_NAME) !== -1) {
        const curDir = path.dirname(id);

        let pureCSSLink = `./css/${styleName}.scss`;
        const cssLink = path.resolve(curDir, pureCSSLink);

        const isExist = fs.existsSync(cssLink);
        if (!isExist) {
          pureCSSLink = '';
        }

        res = res.replace(new RegExp(TIP_STYLE_NAME, 'g'), pureCSSLink);
      }
      return {
        code: res,
        map: null,
      };
    },
  };
}

