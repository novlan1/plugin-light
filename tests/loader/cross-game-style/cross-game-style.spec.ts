import * as fs from 'fs';
import { crossGameStyleLoader } from '../../../src/webpack-loader/cross-game-style/cross-game-style';


jest.mock('fs');
jest.mock('loader-utils', () => ({
  getOptions: jest.fn()
    .mockReturnValueOnce({ platforms: ['mp-weixin'] })
    .mockReturnValueOnce({ styleName: ['gp', 'pvp'], platforms: ['ALL'] })
    .mockReturnValueOnce({ styleName: ['gp', 'pvp'], platforms: ['ALL'] })
    .mockReturnValue({ platforms: ['ALL'] }),
}));
jest.mock('../../../src/webpack-loader/cross-game-style/style-name', () => ({
  getStyleName: jest.fn()
    .mockReturnValueOnce(['gp', 'pvp'])
    .mockReturnValueOnce(['gp', 'pvp'])
    .mockReturnValue('gp'),
}));

async function callLoader(source: string, context = '') {
  let res: string | Buffer | undefined;
  res = crossGameStyleLoader.call({
    async: () => (
      err: Error | undefined | null,
      source?: string | Buffer,
    ) => process.nextTick(() => res = source),
    context,
  }, source);
  await new Promise(process.nextTick);
  return res;
}

describe('crossGameStyleLoader', () => {
  const oldEnv = process.env;

  beforeEach(() => {
    process.env = { ...oldEnv };
  });

  afterEach(() => {
    process.env = oldEnv;
  });

  it('loader not used', async () => {
    const source = '<style scoped lang="scss" src="@TIP_STYLE_NAME">\n</style>';
    expect(await callLoader(source)).toBe(source);
  });

  it('styleName is an array', async () => {
    jest.mocked(fs.existsSync).mockReturnValue(true);
    const source = '<style scoped lang="scss" src="@TIP_STYLE_NAME">\n</style>';
    expect(await callLoader(source)).toMatchSnapshot();
  });

  it('get styleName array from options', async () => {
    jest.mocked(fs.existsSync)
      .mockReturnValueOnce(false)
      .mockReturnValue(true);
    const source = '<style scoped lang="scss" src="@TIP_STYLE_NAME">\n</style>';
    expect(await callLoader(source)).toMatchSnapshot();
  });

  it('loader used but without match', async () => {
    const source = '<style scoped lang="scss" src="./css/gp.scss">\n</style>';
    expect(await callLoader(source)).toBe(source);
  });

  it('inexistent file', async () => {
    jest.mocked(fs.existsSync).mockReturnValue(false);
    const source = '<style scoped lang="scss" src="@TIP_STYLE_NAME">\n</style>';
    expect(await callLoader(source)).toMatchSnapshot();
  });

  it('file exists', async () => {
    jest.mocked(fs.existsSync).mockReturnValue(true);
    const source = '<style scoped lang="scss" src="@TIP_STYLE_NAME">\n</style>';
    expect(await callLoader(source)).toMatchSnapshot();
  });

  it('get styleName from options', async () => {
    jest.mocked(fs.existsSync).mockReturnValue(true);
    const source = '<style scoped lang="scss" src="@TIP_STYLE_NAME">\n</style>';
    expect(await callLoader(source)).toMatchSnapshot();
  });
});
