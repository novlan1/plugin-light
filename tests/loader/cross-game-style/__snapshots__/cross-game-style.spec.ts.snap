// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`crossGameStyleLoader file exists 1`] = `
"<style scoped lang="scss" src="./css/gp.scss">
</style>"
`;

exports[`crossGameStyleLoader get styleName array from options 1`] = `
"<style scoped lang="scss" src="./css/pvp.scss">
</style>"
`;

exports[`crossGameStyleLoader get styleName from options 1`] = `
"<style scoped lang="scss" src="./css/gp.scss">
</style>"
`;

exports[`crossGameStyleLoader inexistent file 1`] = `
"<style scoped lang="scss" src="">
</style>"
`;

exports[`crossGameStyleLoader styleName is an array 1`] = `
"<style scoped lang="scss" src="">
</style><style scoped lang="scss">.gp {@import './css/gp.scss';}.pvp {@import './css/pvp.scss';}</style>"
`;
